### PR TITLE
Poser defaults

### LINKED ISSUE #50: Providing default foreign key values in Laravel factories
Hey, Luke. I've been manually creating class-based factories for a while and can appreciate how great this package is. Really looking forward to using it. I'm wondering what your recommendation is when using default values for foreign keys in your Laravel factories? For example, I typically do something like this.
```
$factory->define(App\Post::class, function (Faker $faker) {
    return [
        'author_id' => factory(Author::class),
        'title' => $faker->title
    ];
});
```
However, when I create a Post using Poser I end up with two Authors in the database.
```
PostFactory::new()
    ->forAuthor(
        Author::new()->withAttributes(['name' => 'Bob'])
    )->create();
```
Do you recommend removing foreign key defaults in Laravel factories? If so, is there an alternative in Poser?

### PR DESCRIPTION
**Pull request by lukeraymonddowning**
After using Poser for a while, you may wish to provide default relationships that should be applied every time we create
a model. For example, in our examples, a `Customer` requires a related `User` model. To facilitate this, Poser understands
a `default[RelationshipType][RelationshipMethodName]` syntax inside Poser Factories. To illustrate:

```php
class CustomerFactory extends Factory
{

    public function defaultForUser()
    {
        return UserFactory::new()->withAttributes(["name" => "Joe Bloggs"]);
    }

}
```

Now, every time we call...

```php
CustomerFactory::new()->create();
```

...a `User` called `Joe Bloggs` will be automatically assigned
to the customer. However, if we call...

```php
CustomerFactory::new()->withUser(UserFactory::new()->withAttributes(["name" => "John Doe"]))->create();
```

...the default will be ignored and instead a `User` called "John Doe" will be assigned to the `Customer`.

For with/has relationship types, the case is much the same:

```php
class UserFactory extends Factory
{

    public function defaultWithAddress() 
    {
        return AddressFactory::new();
    }

    public function defaultHasCustomers()
    {
        return CustomerFactory::times(10);
    }

}
```

In this case, when we call `UserFactory::new()->create()`, it will be given an `Address` and 10 `Customer`s.

If we would like to ignore your defaults for a given test, simply chain the `withoutDefaults` method to your Factory.
call.

```php
UserFactory::new()->withoutDefaults()->create();
```

This will create a `User` with no `Address` or `Customer`s. If you would only like to ignore certain defaults, you may
pass them as properties to the `withoutDefaults` method.

```php
UserFactory::new()->withoutDefaults('customers')->create();
``` 

Now, our created `User` will have no `Customer`s but will have a default `Address`.

### CODE REVIEW

**Comment by lukeraymonddowning** [comment]
@bradroberts @AlanHolmes @andreich1980 

Here is my work on defaults for issue #50 

A lot of changes here so I would appreciate feedback and suggestions 👍

File: README.md [hunk 1]
```diff
@@ -709,6 +715,75 @@ $users = UserFactory::times(10)->create(["name" => "Joe"], ["name" => "Jane"]);
...
+
+```php
+CustomerFactory::new()->create();
+```
+
+...a `User` called `Joe Bloggs` will be automatically assigned
+to the customer. However, if we call...
+
+```php
+CustomerFactory::new()->withUser(UserFactory::new()->withAttributes(["name" => "John Doe"]))->create();
```
Comment (bradroberts) [inline README.md]: Should this be `->forUser`?

File: README.md [hunk 2]
```diff
@@ -722,6 +797,52 @@ When we call `UserFactory::new()->withClients()()`, Poser understands that you'r
...
+```  
+
+Now, we can alter the Customer Poser factory, like so:
+
+```php
+class CustomerFactory extends Factory {
+
+    public function defaultForUser()
+    {
+        return UserFactory::new()->withAttributes(['name' => "John Doe"]);
```
Comment (bradroberts) [inline README.md]: Maybe just `return UserFactory::new()` since the callback this replaces doesn't specify a name attribute?

**Comment by andreich1980** [comment]
Read through the code. Everything seems fine. Created a test for that case.

This PR should also fix a problem described in #42 
Do I have rights to commit to this PR (I'd like to add a test for that)? Or I should create my own new PR for that? @lukeraymonddowning ?

**Review by bradroberts — COMMENTED** [review]
Wow! This looks fantastic! I just had a couple of very nitpicky comments on the readme.

**Comment by lukeraymonddowning** [comment]
> Do I have rights to commit to this PR (I'd like to add a test for that)? Or I should create my own new PR for that? @lukeraymonddowning ?

I think you should have rights here @andreich1980 👍

File: README.md (hunk 1 above)
Comment (lukeraymonddowning) [reply README.md]: You're correct, thanks! Sorted in latest push.

File: README.md (hunk 2 above)
Comment (lukeraymonddowning) [reply README.md]: Good idea, I agree. Pulled across from testing and by this point, my mind was a mess 😁👌

Sorted in latest push.

**Comment by AlanHolmes** [comment]
Code looks good to me 👍 (not had chance to test it)

**Comment by andreich1980** [comment]
Created a PR #55 to this branch with tests

**Comment by andreich1980** [comment]
> Conflicting files
> tests/storage/composer.json
@lukeraymonddowning you can leave only `"autoload"` and throw out `"autoload-dev"` from the `tests/storage/composer.json`
My bad, I commited two versions via two my PRs.

**Comment by lukeraymonddowning** [comment]
And we have liftoff! Thanks all.

**Pull request merged**